routes/dao_coin_exchange.go

```diff
@@ -0,0 +1,254 @@
+package routes
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/deso-protocol/core/lib"
+	"github.com/holiman/uint256"
+	"github.com/pkg/errors"
+	"io"
+	"math/big"
+	"net/http"
+)
+
+type GetDAOCoinLimitOrdersRequest struct {
+	DAOCoin1CreatorPublicKeyBase58CheckOrUsername string `safeForLogging:"true"`
+	DAOCoin2CreatorPublicKeyBase58CheckOrUsername string `safeForLogging:"true"`
+}
+
+type GetDAOCoinLimitOrdersResponse struct {
+	Orders []DAOCoinLimitOrderEntryResponse
+}
+
+type DAOCoinLimitOrderEntryResponse struct {
+	TransactorPublicKeyBase58Check string `safeForLogging:"true"`
+
+	BuyingDAOCoinCreatorPublicKeyBase58Check  string `safeForLogging:"true"`
+	SellingDAOCoinCreatorPublicKeyBase58Check string `safeForLogging:"true"`
+
+	ExchangeRateCoinsToSellPerCoinToBuy float64 `safeForLogging:"true"`
+	QuantityToFill                      float64 `safeForLogging:"true"`
+
+	OperationType DAOCoinLimitOrderOperationTypeString
+
+	OrderID string
+}
+
+func (fes *APIServer) GetDAOCoinLimitOrders(ww http.ResponseWriter, req *http.Request) {
+	decoder := json.NewDecoder(io.LimitReader(req.Body, MaxRequestBodySizeBytes))
+	requestData := GetDAOCoinLimitOrdersRequest{}
+	if err := decoder.Decode(&requestData); err != nil {
+		_AddBadRequestError(
+			ww,
+			fmt.Sprintf("GetDAOCoinLimitOrders: Problem parsing request body: %v", err),
+		)
+		return
+	}
+
+	if requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername == "" &&
+		requestData.DAOCoin2CreatorPublicKeyBase58CheckOrUsername == "" {
+		_AddBadRequestError(
+			ww,
+			fmt.Sprint("GetDAOCoinLimitOrders: Must provide either a "+
+				"DAOCoin1CreatorPublicKeyBase58CheckOrUsername or DAOCoin2CreatorPublicKeyBase58CheckOrUsername "+
+				"or both"),
+		)
+		return
+	}
+
+	utxoView, err := fes.backendServer.GetMempool().GetAugmentedUniversalView()
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetDAOCoinLimitOrders: Problem fetching utxoView: %v", err))
+		return
+	}
+
+	coin1PKID := &lib.ZeroPKID
+	coin2PKID := &lib.ZeroPKID
+
+	coin1ProfilePublicBase58Check := ""
+	coin2ProfilePublicBase58Check := ""
+
+	if requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername != "" {
+		coin1ProfilePublicBase58Check, coin1PKID, err = fes.validateCreatorPublicKeyBase58CheckOrUsername(
+			utxoView,
+			requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername,
+		)
+		if err != nil {
+			_AddBadRequestError(
+				ww,
+				fmt.Sprintf("GetDAOCoinLimitOrders: Invalid DAOCoin1CreatorPublicKeyBase58CheckOrUsername: %v", err),
+			)
+			return
+		}
+	}
+
+	if requestData.DAOCoin2CreatorPublicKeyBase58CheckOrUsername != "" {
+		coin2ProfilePublicBase58Check, coin2PKID, err = fes.validateCreatorPublicKeyBase58CheckOrUsername(
+			utxoView,
+			requestData.DAOCoin2CreatorPublicKeyBase58CheckOrUsername,
+		)
+		if err != nil {
+			_AddBadRequestError(
+				ww,
+				fmt.Sprintf("GetDAOCoinLimitOrders: Invalid DAOCoin2CreatorPublicKeyBase58CheckOrUsername: %v", err),
+			)
+			return
+		}
+	}
```

Comment:
```suggestion
	
	getCoinPKIDAndPubKey := func(creatorPubKeyOrUsername string) (*lib.PKID, string, error) {
		coinPKID := &lib.ZeroPKID
		coinPublicKeyBase58Check := ""
		if creatorPubKeyOrUsername != "" {
			coinPublicKeyBase58Check, coinPKID, err = fes.validateCreatorPublicKeyBase58CheckOrUsername(
				utxoView,
				requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername,
			)
			if err != nil {
				return nil, "", nil
			}
		}
		return coinPKID, coinPublicKeyBase58Check, nil
	}
	
	coin1PKID, coin1ProfilePublicBase58Check, err := getCoinPKIDAndPubKey(requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername)
	if err != nil {
		_AddBadRequestError(
			ww,
			fmt.Sprintf("GetDAOCoinLimitOrders: Invalid DAOCoin1CreatorPublicKeyBase58CheckOrUsername: %v", err),
		)
	}

	coin2PKID, coin2ProfilePublicBase58Check, err := getCoinPKIDAndPubKey(requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername)
	if err != nil {
		_AddBadRequestError(
			ww,
			fmt.Sprintf("GetDAOCoinLimitOrders: Invalid DAOCoin2CreatorPublicKeyBase58CheckOrUsername: %v", err),
		)
	}
```

Reply:
refactor into a locally scoped function

Reply:
Tried making this change locally. But TBH, I think the code reads more clearly as-is given a locally scoped function adds another level of indirection & nesting without deduplicating code any further. If you don't have strong feelings about this, my preference is to leave it as is.

Reply:
Fine by me - don't feel strongly about this